lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -0,0 +1,285 @@
+﻿//===- SCFToCalyx.cpp - SCF to Calyx pass entry point -----------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This is the main SCF to Calyx conversion pass implementation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Conversion/SCFToCalyx/SCFToCalyx.h"
+#include "../PassDetail.h"
+#include "circt/Dialect/Calyx/CalyxOps.h"
+#include "mlir/Conversion/LLVMCommon/ConversionTarget.h"
+#include "mlir/Conversion/LLVMCommon/Pattern.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/AsmState.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include <variant>
+
+using namespace llvm;
+using namespace mlir;
+
+namespace circt {
+
+//===----------------------------------------------------------------------===//
+// Lowering state classes
+//===----------------------------------------------------------------------===//
+
+/// ComponentLoweringState handles the current state of lowering of a Calyx
+/// component. It is mainly used as a key/value store for recording information
+/// during partial lowering, which is required at later lowering passes.
+class ProgramLoweringState;
+class ComponentLoweringState {
+public:
+  ComponentLoweringState(ProgramLoweringState &pls, calyx::ComponentOp compOp)
+      : programLoweringState(pls), compOp(compOp) {}
+
+  ProgramLoweringState &getProgramState() { return programLoweringState; }
+
+  /// Returns a unique name within compOp with the provided prefix.
+  std::string getUniqueName(StringRef prefix) {
+    std::string prefixStr = prefix.str();
+    unsigned idx = prefixIdMap[prefixStr];
+    prefixIdMap[prefixStr]++;
```

Comment:
nit: don't use post-increment unless that's the behavior you want. Pre-increment should be fine here.

Reply:
Right, link to coding standards: https://llvm.org/docs/CodingStandards.html#prefer-preincrement

Reply:
According to the [programmer's manual](https://llvm.org/docs/ProgrammersManual.html#llvm-adt-densemap-h), it seems a DenseMap is more appropriate here?

Reply:
A key difference is for DenseMap, `iterators are invalidated after every insertion`. Perhaps this is an issue for you?

Reply:
`std::string` does not have a `DenseMap` specialization, hence the `std::map`. This also seems to be what is generally used when `std::string`'s are used as the map key.